= Nextflow Implementaton

== PART 1: Data preparation

A first step in any pipeline is to prepare the input data. You will find 
all the data required to run the pipeline in the folder `data` 
within the `ngs2017-nf` repository directory. 

There are four data inputs that we will use in this tutorial:

. *Genome File* (`data/genome.fa`)
* Human chromosome 22 in FASTA file format

. *Read Files* (`data/reads/`)
* Sample 1: 50 bp paired-end reads (`rep1_1.fq.gz` and `rep1_2.fq.gz`).

. *Variants File* (`data/known_variants.vcf.gz`)
* Known variants, gunzipped as a Variant Calling File (VCF) format.

. *Blacklist File* (`data/blacklist.bed`)
* Genomic locations which are known to produce artifacts and spurious variants in Browser Extensible Data (BED) format.


***


=== Getting Started: Specify the input data with parameters.
We can begin writing the pipeline by creating and editing a text file called `main.nf`
from the `ngs2017-nf` repository directory with your favourite text editor. In this example we are using nano:

----
$ nano main.nf
----

We edit can edit this file to specify the default parameters.

----
/*
 * Define the default parameters <1>
 */

params.genome     = "$baseDir/data/genome.fa" <2>
params.variants   = "$baseDir/data/known_variants.vcf.gz"
params.blacklist  = "$baseDir/data/blacklist.bed"
params.reads      = "$baseDir/data/reads/rep1_{1,2}.fq.gz" <3>
params.results    = "results" <4>
params.gatk       = '/usr/local/bin/GenomeAnalysisTK.jar' <5>
----

TIP: With `nano` you can paste the above text into your `main.nf` script with the `shift`+`insert` keys.

<1> The `/\*`, `*` and `*/` specify comment lines which are ignored by Nextflow.

<2> The variable `$baseDir` specifies the base directory of the pipeline, in this case the directory `ngs2017-nf`.

<3> The `reads` parameter uses parameter expansion to specify the forward (`rep1_1.fq.gz`) and reverse (`rep1_2.fq.gz`) reads are pairs of the same sample.

<4> The `results` parameter is used to specify a directory called `results`.

<5> The `gatk` parameter specifies the location of the GATK jar file.


Once you have the default parameters in the `main.nf` file, you can save and run the pipeline for the first time.

TIP: With `nano` you can save and the file with `Ctrl+O`, then `Enter`, followed by `Ctrl+X`.

We run the pipeline with the following command:

----
$ nextflow run main.nf
----

You should see the pipeline launch and then exit.

----
N E X T F L O W  ~  version 0.24.1
Launching `main.nf` [nauseous_wright] - revision: 83a0a5415a
----

***
=== Problem #1
Great, now let get our data into variables and https://www.nextflow.io/docs/latest/channel.html[channels]. Open the `main.nf` file and copy the lines below.

This time you must fill the `BLANK` spaces with the correct constructors.

----
/*
 *  Parse the input parameters
 */

genome_file     = BLANK
variants_file   = BLANK
blacklist_file  = BLANK
reads_ch        = BLANK
GATK            = BLANK
----

TIP: The first three should specify file objects containing a single `file`. For the reads channel, we can use a ttps://www.nextflow.io/docs/latest/channel.html#channel-factory[channel contructor] for pairs of files. The final one, `GATK` is simply a creating a Nextflow variable specifying the relative path of the GATK file and does not need a constructor.


*link:solutions/solution_01.adoc[Problem #1 Solution]*


***

==== Process 1A: Create a FASTA genome index with samtools for GATK

Now we have our inputs set up we can move onto our first process. 

In our first process we will create a genome index using http://www.htslib.org/[samtools]

Copy the code below and paste it at the end of `main.nf`.

=== Problem #2
Your aim is to insert the correct input name from above into
the input step (written as `BLANK`) of the process and run the pipeline.

----
/*
 * Process 1A: Create a FASTA genome index with samtools
 */

process '1A_prepare_genome_samtools' { <1>

  input: 
      file genome from BLANK <2>
      
  output:
      file "${genome}.fai" into genome_index_ch <3>
      
  script:
  """
  samtools faidx ${genome} <4>
  """
}
----

In plain english, the process could be written as:


<1> A **process** called 1A_prepare_genome_samtools

<2> takes as **input** the genome file from `BLANK`

<3> and creates as **output** a genome index file which goes into channel `genome_index_ch`

<4> **script**: using samtools create the genome index from the genome file

*link:solutions/solution_02.adoc[Problem #2 Solution]*

Now we have out first process finished!

***

==== Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK

Our first process created the genome index for GATK using samtools. For the next process we must do something very similar, this time creating a genome sequence dictionary using https://broadinstitute.github.io/picard/[Picard].

=== Problem #3

Fill in the `BLANK` words for both the input and output sections. 

TIP: You can choose any channel output name that makes sense to you.
----
/*
 * Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK
 */

process '1B_prepare_genome_picard' {

  input:
      file genome BLANK BLANK
      
  output:
      file "${genome.baseName}.dict" BLANK BLANK

  script:
  """
  PICARD=`which picard.jar`
  java -jar \$PICARD CreateSequenceDictionary R= $genome O= ${genome.baseName}.dict
  """
}
----

NOTE: `.baseName` returns the filename without the file suffix. If `"${genome}"` is `human.fa`, then `"${genome.baseName}.dict"` is `human.dict`.

*link:solutions/solution_03.adoc[Problem #3 Solution]*


***

==== Process 1C: Create STAR genome index file

Next we must create a genome index for the https://github.com/alexdobin/STAR[STAR] mapping software. 

=== Problem #4

This is a similar exercise except this time both `input` and `output` lines have been left completely `BLANK` and must be completed.

----
/*
 * Process 1C: Create the genome index file for STAR
 */

process '1C_prepare_star_genome_index' {

  input:
      BLANK_LINE
      
  output:
      BLANK_LINE

  script:
  """
  mkdir genome_dir

  STAR --runMode genomeGenerate \
       --genomeDir genome_dir \
       --genomeFastaFiles ${genome} \
       --runThreadN ${task.cpus}
  """
}
----

TIP: The output of the STAR genomeGenerate command is `genome_dir`.

*link:solutions/solution_04.adoc[Problem #4 Solution]*

==== Process 1D: Create a file containing the filtered and recoded set of variants

Next on to something a little more tricky.

The next process takes two inputs: the variants file and the blacklist file.

It should output a channel named `prepared_vcf_ch` which contains a tuple of two files.

NOTE: In Nextflow, tuples can be defined in the input or output using the https://www.nextflow.io/docs/latest/process.html?highlight=set#output-set-of-values[`set`] qualifier.

=== Problem #5

You must fill in the two `BLANK_LINES` in the input and the two `BLANK` output files. 

----
/*
 * Process 1D: Create a file containing the filtered and recoded set of variants
 */

process '1D_prepare_vcf_file' {

  input: 
      BLANK_LINE
      BLANK_LINE

  output:
      set(BLANK,BLANK) into prepared_vcf_ch
  
  script:  
  """
  vcftools --gzvcf $variantsFile -c \ <1>
           --exclude-bed ${blacklisted} \ <2>
           --recode | bgzip -c \
           > ${variantsFile.baseName}.filtered.recode.vcf.gz <3>

  tabix ${variantsFile.baseName}.filtered.recode.vcf.gz <4> 
  """
}
----
<1> The input variable for the variants file
<2> The input variable for the blacklist file 
<3> The first of the two output files
<4> Generates the second output file named `"${variantsFile.baseName}.filtered.recode.vcf.gz.tbi"`

*link:solutions/solution_05.adoc[Problem #5 Solution]*

Congratulations! Part 1 is now complete. 

Try run the pipeline for the project directory with:

```
$ nextflow run main.nf
```

*link:solutions/part1.adoc[You can find the completed pipeline for part 1 here.]*

***



=== PART 2: STAR RNA-Seq Mapping

We have all the data prepared and into channels ready for the more serious steps

==== Process 2: Align RNA-Seq reads to the genome with STAR

In this process, for each sample, we align the reads to our genome using the STAR index we created previously.

The process could be summerised in pseudocode as:

----
    process: 
      2_rnaseq_mapping_star

    synopsis: 
      mapping of the RNA-Seq reads using STAR

    input: 
      the genome fasta file
      the STAR genome index
      a set containing the sample id and paired read files
      
    output: 
      a set containg sample id, aligned bam file & aligned bam file index 
----

=== Problem #6

You must fill in the three `BLANK_LINE` lines in the input and the one `BLANK_LINE` line in the output. 

----
/*
 * Process 2: Align RNA-Seq reads to the genome with STAR
 */

process '2_rnaseq_mapping_star' {

  input: 
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE

  output: 
      BLANK_LINE

  script:    
  """
  # ngs-nf-dev Align reads to genome
  STAR --genomeDir $genomeDir \
       --readFilesIn $reads \
       --runThreadN ${task.cpus} \
       --readFilesCommand zcat \
       --outFilterType BySJout \
       --alignSJoverhangMin 8 \
       --alignSJDBoverhangMin 1 \
       --outFilterMismatchNmax 999
    
  # 2nd pass (improve alignmets using table of splice junctions and create a new index)  
  mkdir genomeDir  
  STAR --runMode genomeGenerate \
       --genomeDir genomeDir \
       --genomeFastaFiles $genome \
       --sjdbFileChrStartEnd SJ.out.tab \
       --sjdbOverhang 75 \
       --runThreadN ${task.cpus}  
    
  # Final read alignments  
  STAR --genomeDir genomeDir \
       --readFilesIn $reads \
       --runThreadN ${task.cpus} \
       --readFilesCommand zcat \
       --outFilterType BySJout \
       --alignSJoverhangMin 8 \
       --alignSJDBoverhangMin 1 \
       --outFilterMismatchNmax 999 \
       --outSAMtype BAM SortedByCoordinate \
       --outSAMattrRGline ID:$replicateId LB:library PL:illumina PU:machine SM:GM12878

  # Index the BAM file
  samtools index Aligned.sortedByCoord.out.bam
  """
}
----

TIP: The final command produces an bam index which is the file name with `.bai` attached.

*link:solutions/solution_06.adoc[Problem #6 Solution]*


== PART 3: GATK Prepare Mapped Reads

Part 3 is a filtering step using GATK.

For each sample, we split all the reads that contain
N characters in their http://genome.sph.umich.edu/wiki/SAM#What_is_a_CIGAR.3F[CIGAR] string.

=== Process 3: Split reads that contain Ns in their CIGAR string

The process creates k+1 new reads (where k is the number of N cigar elements) 
that correspond to the segments of the original read beside/between 
the splicing events represented by the Ns in the original CIGAR.

The process could be summerised in pseudocode as:

----
    process: 
      3_rnaseq_gatk_splitNcigar

    synopsis: 
      split reads on N in CIGAR using GATK

    input: 
      the genome fasta file
      the genome index made with samtools
      the genome dictionary made with picard
      a set containg sample id, aligned bam file and aligned bam file index from the STAR mapping
      
    output: 
      a set containing the sample id, the split bam file and the split bam index file
      
----

=== Problem #7

You must fill in the four `BLANK_LINE` lines in the input and the one `BLANK_LINE` line in the output. 

CAUTION: There is an optional https://www.nextflow.io/docs/latest/process.html#tag[`tag`] line added 
to the start of this process. The https://www.nextflow.io/docs/latest/process.html#tag[`tag`] line allows you to assign a name to a specific 
task (single execution of a process). This is particularly useful when there are many samples all 
using the same process.
 
----
process '3_rnaseq_gatk_splitNcigar' {
  tag OPTIONAL_BLANK
  
  input: 
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE

  output:
      BLANK_LINE
  
  script:
  """
  # SplitNCigarReads and reassign mapping qualities
  java -jar $GATK -T SplitNCigarReads \
                  -R $genome -I $bam \
                  -o split.bam \
                  -rf ReassignOneMappingQuality \
                  -RMQF 255 -RMQT 60 \
                  -U ALLOW_N_CIGAR_READS \
                  --fix_misencoded_quality_scores
  """
}
----

TIP: The GATK command above automatically creates a bam index (.bai) of the split.bam output file

*link:solutions/solution_07.adoc[Problem #7 Solution]*

***

== PART 4: GATK Base Quality Score Recalibration Workflow

Part 4 is a recalibration step using GATK.

=== Process 4: Base recalibrate to detect systematic errors in base quality scores, select unique alignments and index

=== Problem #8




== PART 5: GATK Variant Calling

=== Process 5: Call variants with GATK HaplotypeCaller

== PART 6: Post-process variants file and prepare for Allele-Specific Expression and RNA Editing Analysis

=== Process 6A: Post-process the VCF result

=== Process 6B: Prepare variants file for allele specific expression (ASE) analysis

=== Process 6C: Allele-Specific Expression analysis with GATK ASEReadCounter
