<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<!--[if IE]><meta http-equiv="X-UA-Compatible" content="IE=edge"><![endif]-->
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 1.5.4">
<title>NGS 2017 Workshop</title>
<link rel="stylesheet" href="./css/crg.css">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.5.0/css/font-awesome.min.css">
</head>
<body class="article toc2 toc-left">
<div id="header">
<h1>NGS 2017 Workshop <span class="image" style="float: right"><img src="./assets//NGS_LOGO.png" alt="Logo" width="120" height="120"></span></h1>
<div id="toc" class="toc2">
<div id="toctitle">Table of Contents</div>
<ul class="sectlevel1">
<li><a href="#_nextflow">1. Nextflow</a>
<ul class="sectlevel2">
<li><a href="#_installing_nextflow">1.1. Installing Nextflow</a></li>
<li><a href="#_basic_script">1.2. Basic script</a></li>
<li><a href="#_using_parameters">1.3. Using parameters</a></li>
<li><a href="#_parallel_execution">1.4. Parallel execution</a></li>
<li><a href="#_collect_results">1.5. Collect results</a></li>
<li><a href="#_use_a_computing_cluster">1.6. Use a computing cluster</a></li>
<li><a href="#_resume_pipeline_execution">1.7. Resume pipeline execution</a></li>
<li><a href="#_integration_with_github">1.8. Integration with GitHub</a></li>
<li><a href="#_docker_containers">1.9. Docker containers</a></li>
</ul>
</li>
<li><a href="#_worflow_description">2. Worflow Description</a>
<ul class="sectlevel2">
<li><a href="#_2_pass_mapping_to_the_reference">2.1. 2-pass mapping to the reference</a></li>
<li><a href="#_split_n_trim_and_reassign_mapping_qualities">2.2. Split&#8217;N&#8217;Trim and reassign mapping qualities</a></li>
<li><a href="#_base_recalibration">2.3. Base Recalibration</a></li>
<li><a href="#_variant_calling">2.4. Variant Calling</a></li>
<li><a href="#_variant_filtering">2.5. Variant Filtering</a></li>
<li><a href="#_variant_post_processing">2.6. Variant Post-processing</a></li>
</ul>
</li>
<li><a href="#_nextflow_implementaton">3. Nextflow Implementaton</a>
<ul class="sectlevel2">
<li><a href="#_part_1_data_preparation">3.1. PART 1: Data preparation</a>
<ul class="sectlevel3">
<li><a href="#_specify_the_input_data_with_parameters">3.1.1. Specify the input data with parameters.</a></li>
<li><a href="#_process_1a_create_a_fasta_genome_index_code_fai_code_with_samtools_for_gatk">3.1.2. Process 1A: Create a FASTA genome index (<code>.fai</code>) with samtools for GATK</a></li>
<li><a href="#_process_1b_create_a_fasta_genome_sequence_dictionary_with_picard_for_gatk">3.1.3. Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK</a></li>
<li><a href="#_process_1c_create_star_genome_index_file">3.1.4. Process 1C: Create STAR genome index file</a></li>
<li><a href="#_process_1d_create_a_file_containing_the_filtered_and_recoded_set_of_variants">3.1.5. Process 1D: Create a file containing the filtered and recoded set of variants</a></li>
</ul>
</li>
<li><a href="#_part_2_star_rna_seq_mapping">3.2. PART 2: STAR RNA-Seq Mapping</a>
<ul class="sectlevel3">
<li><a href="#_process_2_align_rna_seq_reads_to_the_genome_with_star">3.2.1. Process 2: Align RNA-Seq reads to the genome with STAR</a></li>
</ul>
</li>
<li><a href="#_part_3_gatk_prepare_mapped_reads">3.3. PART 3: GATK Prepare Mapped Reads</a>
<ul class="sectlevel3">
<li><a href="#_process_3_split_reads_that_contain_ns_in_their_cigar_string">3.3.1. Process 3: Split reads that contain Ns in their CIGAR string</a></li>
</ul>
</li>
<li><a href="#_part_4_gatk_base_quality_score_recalibration_workflow">3.4. PART 4: GATK Base Quality Score Recalibration Workflow</a>
<ul class="sectlevel3">
<li><a href="#_process_4_base_recalibrate_to_detect_systematic_errors_in_base_quality_scores_select_unique_alignments_and_index">3.4.1. Process 4: Base recalibrate to detect systematic errors in base quality scores, select unique alignments and index</a></li>
</ul>
</li>
<li><a href="#_part_5_gatk_variant_calling">3.5. PART 5: GATK Variant Calling</a>
<ul class="sectlevel3">
<li><a href="#_process_5_call_variants_with_gatk_haplotypecaller">3.5.1. Process 5: Call variants with GATK HaplotypeCaller</a></li>
</ul>
</li>
<li><a href="#_part_6_post_process_variants_file_and_prepare_for_allele_specific_expression_and_rna_editing_analysis">3.6. PART 6: Post-process variants file and prepare for Allele-Specific Expression and RNA Editing Analysis</a>
<ul class="sectlevel3">
<li><a href="#_process_6a_post_process_the_vcf_result">3.6.1. Process 6A: Post-process the VCF result</a></li>
<li><a href="#_process_6b_prepare_variants_file_for_allele_specific_expression_ase_analysis">3.6.2. Process 6B: Prepare variants file for allele specific expression (ASE) analysis</a></li>
<li><a href="#_process_6c_allele_specific_expression_analysis_with_gatk_asereadcounter">3.6.3. Process 6C: Allele-Specific Expression analysis with GATK ASEReadCounter</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#_scale_your_nextflow_pipeline_in_a_large_cluster_or_the_cloud">4. Scale your Nextflow pipeline in a large cluster or the cloud</a></li>
</ul>
</div>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph">
<p>This workshop will introduce the best practices proposed to tackle the problem of irreproducible NGS data analyses and allow to make large omics workflows portable across HPC clusters and cloud environments.</p>
</div>
<div class="paragraph">
<p>The practical session will go through the step by step implementation of a pipeline for standard <span class="crg">Variant Calling Analyses with RNA-seq data</span>. For this purpose attendess wiil use Nextflow, a framework for computational pipelines that simplifies the development and deployment of NGS pipelines in a portable manner across different execution environments.</p>
</div>
<div class="admonitionblock warning">
<table>
<tr>
<td class="icon">
<i class="fa icon-warning" title="Warning"></i>
</td>
<td class="content">
Add dataset description
</td>
</tr>
</table>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
A cheatsheet with basic Linux and Cluster commands is available <a href="cheatsheet.html">here</a>.
</td>
</tr>
</table>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_nextflow">1. Nextflow</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Nextflow is a framework that simplifies writing complex parallel computational
pipelines in a portable and reproducible manner. Parallelization is automatically managed
by the framework and it is implicitly defined by the processes input and output declarations.</p>
</div>
<div class="paragraph">
<p>The built-in support for <a href="https://www.docker.com">Docker</a> containers technology and the <a href="https://www.github.com">Github</a>
sharing platform enables pipelines to be deployed, along with all their dependencies, across
multiple platforms without any modifications, making it possible to share them and replicate
their results in a predictable manner.</p>
</div>
<div class="sect2">
<h3 id="_installing_nextflow">1.1. Installing Nextflow</h3>
<div class="paragraph">
<p>Nextflow is a command line tool. It only requires a Unix-like operating system and a Java 7/8 available
in the running environment.</p>
</div>
<div class="paragraph">
<p>It can be installed with the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">curl -fsSL get.nextflow.io | bash
mv nextflow ~/bin</code></pre>
</div>
</div>
<div class="paragraph">
<p>To check the installed version use the command below:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow info</code></pre>
</div>
</div>
<div class="listingblock">
<div class="content">
<pre>  Version: 0.17.3 build 3495
  Modified: 18-02-2016 11:00 UTC (12:00 CEST)
  System: Mac OS X 10.10.5
  Runtime: Groovy 2.4.5 on Java HotSpot(TM) 64-Bit Server VM 1.8.0_40-b27
  Encoding: UTF-8 (UTF-8)</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_basic_script">1.2. Basic script</h3>
<div class="paragraph">
<p>Write a basic script to count the number of transcripts and exons in a gene annotation file.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code>#!/usr/bin/env nextflow
echo true
annotation = Channel.fromPath('/users/ngs00/refs/mm65.long.ok.gtf')

process count {

  input:
  file 'annotation.gtf' from annotation

  '''
  grep Sec23a annotation.gtf &gt; Sec23a.gff
  awk '$3=="transcript"' Sec23a.gff | wc -l
  awk '$3=="exon"' Sec23a.gff | wc -l
  '''
}</code></pre>
</div>
</div>
<div class="paragraph">
<p>This examples introduces the basic syntax of a Nextflow process and shows how to execute
any existing piece of code or script available in the hosting environment.</p>
</div>
<div class="paragraph">
<p>Save the code showed above in a file named <code>tutorial1.nf</code>, then run it using the following
command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow run tutorial1.nf</code></pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_using_parameters">1.3. Using parameters</h3>
<div class="paragraph">
<p>Parameters allow the same script to be used specifying different input values.</p>
</div>
<div class="paragraph">
<p>By convention Nextflow parameters are defined at the top of the script file
providing a default value for each of them.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code>#!/usr/bin/env nextflow
echo true

params.gene = 'Sec23a'
params.annot = '/users/ngs00/refs/mm65.long.ok.gtf'

annotation = Channel.fromPath(params.annot)

process count {
  input:
  file 'annotation.gtf' from annotation

  shell:
  '''
  grep !{params.gene} annotation.gtf &gt; gff
  awk '$3=="transcript"' gff | wc -l
  awk '$3=="exon"' gff | wc -l
  '''
}</code></pre>
</div>
</div>
<div class="paragraph">
<p>The actual parameter value can be provided on the launch command line as an extra option
prefixing the parameter name with <code>--</code> and then specifying a value e.g. <code>--gene Sec1</code>.</p>
</div>
<div class="paragraph">
<p>Save the above code in a file named <code>tutorial2.nf</code>, then execute it providing a different
gene name or an annotation file as shown below:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow run tutorial2.nf --gene Sec1
:
nextflow run tutorial2.nf --annot /users/ngs00/refs/gencode.vM7.annotation.gtf --gene Sec23b</code></pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_parallel_execution">1.4. Parallel execution</h3>
<div class="paragraph">
<p>Nextflow processes are implicitly executed in a parallel manner whenever multiple inputs
are provided. The following example executes a parallel task for each pair of gene name and annotation
file specified.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code>#!/usr/bin/env nextflow
echo true

params.gene = 'Sec23a'
params.annot = '/users/ngs00/refs/mm65.long.ok.gtf'

annotation = Channel.fromPath(params.annot)
genes = params.gene.tokenize(', ')

process count {
  input:
  file 'annotation.gtf' from annotation
  each gene from genes

  shell:
  '''
  echo !{gene}
  grep !{gene} annotation.gtf &gt; gff
  awk '$3=="transcript"' gff | wc -l
  awk '$3=="exon"' gff | wc -l
  '''
}</code></pre>
</div>
</div>
<div class="paragraph">
<p>Save the above content to a file named <code>tutorial3.nf</code>, then run it proving more than an
annotation file as shown below:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow run tutorial3.nf --annot '/users/ngs00/refs/*.gtf'
:
nextflow run tutorial3.nf --gene Sec23a,Tulp1,Trex1
:</code></pre>
</div>
</div>
<div class="paragraph">
<p>Each of file matching the specified glob pattern will trigger the execution
of a parallel <code>count</code> process. The same happens when specifying multiple gene names.</p>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
The <code>each</code> clause results in the execution of a parallel task for all the combinations
of annotation files with the specified gene names.
</td>
</tr>
</table>
</div>
</div>
<div class="sect2">
<h3 id="_collect_results">1.5. Collect results</h3>
<div class="paragraph">
<p>This example shows how collect the output produced by multiple parallel processes
into a file and prints the resulting content.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code>#!/usr/bin/env nextflow

params.gene = 'Sec23a'
params.annot = '/users/ngs00/refs/mm65.long.ok.gtf'

annotation = Channel.fromPath(params.annot)
genes = params.gene.tokenize(', ')

process count {
  input:
  each gene from genes
  file annot from annotation

  output:
  stdout into result

  shell:
  '''
  echo !{annot.baseName}
  echo !{gene}
  grep !{gene} !{annot} &gt; gff
  awk '$3=="transcript"' gff | wc -l
  awk '$3=="exon"' gff | wc -l
  '''
}

result
    .map { str -&gt; str.readLines().join(',') }  <i class="conum" data-value="1"></i><b>(1)</b>
    .collectFile(newLine: true)  <i class="conum" data-value="2"></i><b>(2)</b>
    .println { it.text }  <i class="conum" data-value="3"></i><b>(3)</b></code></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>The <code>map</code> operator transforms the multi-lines output into a comma-separated string.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>The <code>collectFile</code> operator gathers the produced strings and append them into a file.</td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td>The <code>println</code> operator prints the file content.</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>Save the above script to a file named <code>tutorial4.nf</code>, then run it by using the
following command in your shell terminal:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow run tutorial4.nf --annot '/users/ngs00/refs/*.gtf' --gene Sec23a,Tulp1</code></pre>
</div>
</div>
<div class="paragraph">
<p>It will output a text similar to the one below following:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>gencode.vM9.annotation,Tulp1,12,71
gencode.vM7.annotation,Sec23a,5,47
mm65.long.ok,Sec23a,5,47
gencode.v22.annotation,Sec23a,0,0
gencode.v22.annotation,Tulp1,0,0
gencode.vM7.annotation,Tulp1,12,71
mm65.long.ok,Tulp1,12,71
dmel-all-no-analysis-r6.05,Tulp1,0,0
dmel-all-no-analysis-r6.05,Sec23a,0,0
gencode.vM9.annotation,Sec23a,5,47</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_use_a_computing_cluster">1.6. Use a computing cluster</h3>
<div class="paragraph">
<p>When a pipeline runs many computing intensive tasks a batch scheduler is required
to submit the job executions to a cluster of computers.</p>
</div>
<div class="paragraph">
<p>Nextflow manages the execution with the batch scheduler in a transparent manner
without any change in the pipeline code. It only requires a few settings in the
pipeline configuration file:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code>process {
    executor = 'sge'
    queue = 'NGS'
    memory = '1 GB'
}</code></pre>
</div>
</div>
<div class="paragraph">
<p>Save the content showed above in a file named <code>nextflow.config</code>, then launch
the script execution as before:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow run tutorial4.nf -bg --annot '/users/ngs00/refs/*.gtf' --gene Sec23a,Tulp1,Trex &gt; log</code></pre>
</div>
</div>
<div class="paragraph">
<p>You can check tasks are submitted to the cluster using the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">qstat</code></pre>
</div>
</div>
<div class="paragraph">
<p>The following platforms are currently supported:</p>
</div>
<div class="ulist">
<ul>
<li>
<p>Sun/Open Grid Engine</p>
</li>
<li>
<p>Univa Grid Engine</p>
</li>
<li>
<p>Linux SLURM</p>
</li>
<li>
<p>IBM LSF</p>
</li>
<li>
<p>Torque/PBS</p>
</li>
</ul>
</div>
</div>
<div class="sect2">
<h3 id="_resume_pipeline_execution">1.7. Resume pipeline execution</h3>
<div class="paragraph">
<p>Launch the script execution as shown below:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow run tutorial4.nf --annot '/users/ngs00/refs/*.gtf' --gene Sec23a -resume</code></pre>
</div>
</div>
<div class="paragraph">
<p>The <code>-resume</code> command line option will force to skip the execution of tasks that have
been already executed successfully.</p>
</div>
<div class="paragraph">
<p>It will print an output similar to the following:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>N E X T F L O W  ~  version 0.17.3
Launching tutorial4.nf
[warm up] executor &gt; sge
[85/145369] Cached process &gt; count (2)
[35/054b18] Cached process &gt; count (1)
[4a/36a5d1] Cached process &gt; count (3)
gencode.vM7.annotation,Sec23a,5,47
mm65.long.ok,Sec23a,5,47</pre>
</div>
</div>
<div class="paragraph">
<p>If you add other genes by using the <code>--gene</code> option only the tasks required by the new input
will be executed. For example:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow run tutorial4.nf --annot '/users/ngs00/refs/*.gtf' --gene Sec23a,Tulp2,Trex2 -resume</code></pre>
</div>
</div>
<div class="paragraph">
<p>Only the tasks for which a new input is specified will be executed.</p>
</div>
</div>
<div class="sect2">
<h3 id="_integration_with_github">1.8. Integration with GitHub</h3>
<div class="paragraph">
<p>Nextflow provides built-in support for the <a href="https://git-scm.com/">Git</a> tool and the <a href="https://github.com">GitHub</a> source code management
plaftorm. This makes it possible to share and deploy a pipeline hosted in the GitHub
platform with ease.</p>
</div>
<div class="paragraph">
<p>For the sake of this tutorial we will use <a href="https://github.com/nextflow-io/rnatoy">RNA-Toy</a>,
a proof of concept RNA-Seq pipeline implemented with Nextflow.</p>
</div>
<div class="paragraph">
<p>The pipeline uses the following tools:</p>
</div>
<div class="ulist">
<ul>
<li>
<p><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a></p>
</li>
<li>
<p><a href="https://ccb.jhu.edu/software/tophat/index.shtml">TopHat2</a></p>
</li>
<li>
<p><a href="http://cole-trapnell-lab.github.io/cufflinks/">Cufflinks</a></p>
</li>
</ul>
</div>
<div class="paragraph">
<p>In order to setup the required pipeline dependencies we will use <a href="http://modules.sourceforge.net/">Environment Modules</a>.</p>
</div>
<div class="paragraph">
<p>Add the following setting in your <code>nextflow.config</code> file:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>process {
  module = 'Boost/1.55.0-goolf-1.4.10-no-OFED-Python-2.7.6:Python/2.7.8-goolf-1.4.10-no-OFED:Bowtie2/2.2.8-goolf-1.4.10-no-OFED:TopHat/2.1.0-goolf-1.4.10-no-OFED:Cufflinks/2.2.1-goolf-1.4.10-no-OFED'
}</pre>
</div>
</div>
<div class="paragraph">
<p>Launch the pipeline execution using the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow run nextflow-io/rnatoy</code></pre>
</div>
</div>
<div class="paragraph">
<p>The above command will automatically pull the pipeline from the GitHub repository and run it
using a dataset included in the pipeline itself.</p>
</div>
<div class="paragraph">
<p>Users can provide its own dataset specifying it on the launch command line.</p>
</div>
</div>
<div class="sect2">
<h3 id="_docker_containers">1.9. Docker containers</h3>
<div class="paragraph">
<p>In the previous example we still needed to configure one by one the tools used by the pipeline
by loading the corresponding environment modules.</p>
</div>
<div class="paragraph">
<p>Nextflow provides built-in support for <a href="https://www.docker.com/">Docker</a> containers that
allows binary dependencies to be deployed automatically when the pipeline is executed.</p>
</div>
<div class="paragraph">
<p>Having the Docker engine installed, the previous example can be executed simply using
the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd" data-lang="cmd">nextflow run nextflow-io/rnatoy -with-docker</code></pre>
</div>
</div>
<div class="paragraph">
<p>The <code>-with-docker</code> command line option instructs Nextflow to pull the Docker image defined
in the pipeline execution file. The image contains all the binary dependencies required to
run the pipeline script (i.e. Bowtie, TopHat, Cufflinks, etc.) thus it is not need
to configure them manually.</p>
</div>
<div class="paragraph">
<p>This greatly simplify the pipeline deployment process and guarantee consistent results over time and
across different platforms.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_worflow_description">2. Worflow Description</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_2_pass_mapping_to_the_reference">2.1. 2-pass mapping to the reference</h3>

</div>
<div class="sect2">
<h3 id="_split_n_trim_and_reassign_mapping_qualities">2.2. Split&#8217;N&#8217;Trim and reassign mapping qualities</h3>

</div>
<div class="sect2">
<h3 id="_base_recalibration">2.3. Base Recalibration</h3>

</div>
<div class="sect2">
<h3 id="_variant_calling">2.4. Variant Calling</h3>

</div>
<div class="sect2">
<h3 id="_variant_filtering">2.5. Variant Filtering</h3>

</div>
<div class="sect2">
<h3 id="_variant_post_processing">2.6. Variant Post-processing</h3>

</div>
</div>
</div>
<div class="sect1">
<h2 id="_nextflow_implementaton">3. Nextflow Implementaton</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_part_1_data_preparation">3.1. PART 1: Data preparation</h3>
<div class="paragraph">
<p>A first step in any pipleine is to prepare the input data.</p>
</div>
<div class="paragraph">
<p>You will find all the data required to run the pipeline in the
folder <code>data</code> within the <code>ngs2017-nf</code> repository.</p>
</div>
<div class="paragraph">
<p>There are 4 data inputs that we will use in this tutorial:</p>
</div>
<div class="ulist">
<div class="title">Genome File (<code>data/genome.fa</code>)</div>
<ul>
<li>
<p>Human chromosome 22 in FASTA file format</p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Read Files (<code>data/reads/</code>)</div>
<ul>
<li>
<p>Sample 1: 50 bp paired-end reads (<code>rep1_1.fq.gz</code> and <code>rep1_2.fq.gz</code>).</p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Variants File (<code>data/known_variants.vcf.gz</code>)</div>
<ul>
<li>
<p>Known variants, gunzipped as a Variant Calling File (VCF) format.</p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Blacklist File (<code>data/blacklist.bed</code>)</div>
<ul>
<li>
<p>Genomic locations which are known to produce artifacts and spurious variants in Browser Extensible Data (BED) format.</p>
</li>
</ul>
</div>
<div class="sect3">
<h4 id="_specify_the_input_data_with_parameters">3.1.1. Specify the input data with parameters.</h4>
<div class="paragraph">
<p>We can begin writing the pipeline by creating and editing a text file called <code>main.nf</code>
from the <code>ngs2017-nf</code> repository directory with your favourite text editor. In this example we are using nano:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>nano main.nf</pre>
</div>
</div>
<div class="paragraph">
<p>From here we edit can edit this file to specify the default parameters.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>/*
 * Define the default parameters
 */

params.genome     = "$baseDir/data/genome.fa"
params.variants   = "$baseDir/data/known_variants.vcf.gz"
params.blacklist  = "$baseDir/data/blacklist.bed"
params.reads      = "$baseDir/data/reads/rep1_{1,2}.fq.gz"
params.results    = "results"
params.gatk       = '/usr/local/bin/GenomeAnalysisTK.jar'</pre>
</div>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
With <code>nano</code> you can paste the above text into your <code>main.nf</code> script with the <code>shift</code>+<code>insert</code> keys.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>The <code>\/<strong></code> and <code>\</strong>/</code> specify comment lines which are ignored by Nextflow.</p>
</div>
<div class="paragraph">
<p>The varible <code>$baseDir</code> specifies the base directory of the pipeline, in this case the directory ngs2017-nf.</p>
</div>
<div class="paragraph">
<p>Each of the parameter words specified after the <code>params.</code> can also be
specified at run time on the command line using <code>--</code> followed by th word.
For example, if you wished to run the pipeline with a different genome file,
you could run:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>nextflow run main.nf --genome alternative_genome.fa</pre>
</div>
</div>
<div class="paragraph">
<p>The <code>reads</code> parameter uses parameter expansion to specify the forward (<code>rep1_1.fq.gz</code>) and reverse (<code>rep1_2.fq.gz</code>) reads are pairs of the same sample.</p>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
<div class="title">Parameter expansion</div>
See <a href="http://https://www.nextflow.io/docs/latest/getstarted.html">here</a> for
a more detailed example of how parameter expansion can be used.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>The <code>results</code> parameter is used to specify a directory called <code>results</code>.</p>
</div>
<div class="paragraph">
<p>The <code>gatk</code> parameter specifies the location of the GATK file.</p>
</div>
<div class="paragraph">
<p>Once you have the default parameters in the <code>main.nf</code> file, you can save and run the pipeline for the first time.</p>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
With <code>nano</code>, you can save and the file with <code>Ctrl+O</code> then <code>Enter</code> followed by <code>Ctrl+X</code>.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>We run the pipeline with the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>nextflow run main.nf</pre>
</div>
</div>
<div class="paragraph">
<p>You see should the pipeline launch and then exit.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>$ nextflow run main.nf
N E X T F L O W  ~  version 0.24.1
Launching `main.nf` [nauseous_wright] - revision: 83a0a5415a</pre>
</div>
</div>
<div class="paragraph">
<p>Great, now lets start using the data by putting them into channels.</p>
</div>
<div class="paragraph">
<div class="title">Problem #1</div>
<p>Open the <code>main.nf</code> file again and this time you must fill the <code>BLANK</code> spaces with
the correct channel contructors.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>/*
 *  Parse the input parameters
 */

genome_file     = BLANK
variants_file   = BLANK
blacklist_file  = BLANK
reads_ch        = BLANK
GATK            = BLANK</pre>
</div>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
The first three specify channels containing a single <code>file</code>. For the reads channel, we can use a [special contructor] for pairs. The final one, <code>GATK</code> is simply a creating a Nextflow string specifying the relative path of the GATK file.
</td>
</tr>
</table>
</div>
<div class="listingblock">
<div class="title">Problem #1 Solution</div>
<div class="content">
<pre>/*
 *  Parse the input parameters
 */

genome_file     = file(params.genome)
variants_file   = file(params.variants)
blacklist_file  = file(params.blacklist)
reads_ch        = Channel.fromFilePairs(params.reads)
GATK            = params.gatk</pre>
</div>
</div>
<div class="paragraph">
<p>Now we have our input channels set up we can move onto our first process.</p>
</div>
</div>
<div class="sect3">
<h4 id="_process_1a_create_a_fasta_genome_index_code_fai_code_with_samtools_for_gatk">3.1.2. Process 1A: Create a FASTA genome index (<code>.fai</code>) with samtools for GATK</h4>
<div class="paragraph">
<p>Processes are defined using <code>process</code> followed by the process name.</p>
</div>
<div class="paragraph">
<p>Processes form the heart of Nextflow pipelines and are linked together by channels.</p>
</div>
<div class="paragraph">
<div class="title">Problem #2,</div>
<p>Your aim is to insert the correct input channel name from above into
the input step (written as BLANK) of the process and run the pipeline.</p>
</div>
<div class="listingblock">
<div class="content">
<pre> *
 * Process 1A: Create a FASTA genome index (.fai) with samtools for GATK
 */

process '1A_prepare_genome_samtools' {
  tag "$genome.baseName"

  input:
      file genome from BLANK

  output:
      file "${genome}.fai" into genome_index_ch

  script:
  """
  samtools faidx ${genome}
  """
}</pre>
</div>
</div>
<div class="paragraph">
<p>We can see that the process is named <code>1A_prepare_genome_samtools</code>.</p>
</div>
<div class="paragraph">
<p>It takes a file as input, which becomes referenced by the <code>${genome}</code>
variable in the script section.</p>
</div>
<div class="paragraph">
<p>The output is file called ${genome}.fai (for FASTA index) which becomes
part of a new channel called <code>genome_index_ch</code>.</p>
</div>
<div class="paragraph">
<p>The script section calls for the execution of the <code>samtools</code> program with the
argument <code>faidx</code> for fasta index and the ${genome} as the input.</p>
</div>
<div class="listingblock">
<div class="title">Solution #2,</div>
<div class="content">
<pre> *
 * Process 1A: Create a FASTA genome index (.fai) with samtools for GATK
 */

process '1A_prepare_genome_samtools' {
  tag "$genome.baseName"

  input:
      file genome from genome_file

  output:
      file "${genome}.fai" into genome_index_ch

  script:
  """
  samtools faidx ${genome}
  """
}</pre>
</div>
</div>
<div class="paragraph">
<p>We can see above that the solution is to take the genome file from the
<code>genome_file</code> channel we created.</p>
</div>
<div class="paragraph">
<p>Next step is to</p>
</div>
</div>
<div class="sect3">
<h4 id="_process_1b_create_a_fasta_genome_sequence_dictionary_with_picard_for_gatk">3.1.3. Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK</h4>

</div>
<div class="sect3">
<h4 id="_process_1c_create_star_genome_index_file">3.1.4. Process 1C: Create STAR genome index file</h4>

</div>
<div class="sect3">
<h4 id="_process_1d_create_a_file_containing_the_filtered_and_recoded_set_of_variants">3.1.5. Process 1D: Create a file containing the filtered and recoded set of variants</h4>

</div>
</div>
<div class="sect2">
<h3 id="_part_2_star_rna_seq_mapping">3.2. PART 2: STAR RNA-Seq Mapping</h3>
<div class="sect3">
<h4 id="_process_2_align_rna_seq_reads_to_the_genome_with_star">3.2.1. Process 2: Align RNA-Seq reads to the genome with STAR</h4>

</div>
</div>
<div class="sect2">
<h3 id="_part_3_gatk_prepare_mapped_reads">3.3. PART 3: GATK Prepare Mapped Reads</h3>
<div class="sect3">
<h4 id="_process_3_split_reads_that_contain_ns_in_their_cigar_string">3.3.1. Process 3: Split reads that contain Ns in their CIGAR string</h4>

</div>
</div>
<div class="sect2">
<h3 id="_part_4_gatk_base_quality_score_recalibration_workflow">3.4. PART 4: GATK Base Quality Score Recalibration Workflow</h3>
<div class="sect3">
<h4 id="_process_4_base_recalibrate_to_detect_systematic_errors_in_base_quality_scores_select_unique_alignments_and_index">3.4.1. Process 4: Base recalibrate to detect systematic errors in base quality scores, select unique alignments and index</h4>

</div>
</div>
<div class="sect2">
<h3 id="_part_5_gatk_variant_calling">3.5. PART 5: GATK Variant Calling</h3>
<div class="sect3">
<h4 id="_process_5_call_variants_with_gatk_haplotypecaller">3.5.1. Process 5: Call variants with GATK HaplotypeCaller</h4>

</div>
</div>
<div class="sect2">
<h3 id="_part_6_post_process_variants_file_and_prepare_for_allele_specific_expression_and_rna_editing_analysis">3.6. PART 6: Post-process variants file and prepare for Allele-Specific Expression and RNA Editing Analysis</h3>
<div class="sect3">
<h4 id="_process_6a_post_process_the_vcf_result">3.6.1. Process 6A: Post-process the VCF result</h4>

</div>
<div class="sect3">
<h4 id="_process_6b_prepare_variants_file_for_allele_specific_expression_ase_analysis">3.6.2. Process 6B: Prepare variants file for allele specific expression (ASE) analysis</h4>

</div>
<div class="sect3">
<h4 id="_process_6c_allele_specific_expression_analysis_with_gatk_asereadcounter">3.6.3. Process 6C: Allele-Specific Expression analysis with GATK ASEReadCounter</h4>

</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_scale_your_nextflow_pipeline_in_a_large_cluster_or_the_cloud">4. Scale your Nextflow pipeline in a large cluster or the cloud</h2>
<div class="sectionbody">

</div>
</div>
</div>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.9.1/styles/github.min.css">
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.9.1/highlight.min.js"></script>
<script>hljs.initHighlighting()</script>
</body>
</html>