
= Results overview

For each sample the pipeline creates a folder with sample ID name to write results, and produce a number of files. 

Result files for this workshop can be found in the folder with name `ENCSR000COQ/`. 

There you should see following files:


Variant calls::

* *_final.vcf_*

This file contains all somatic variants (SNVs) called from RNAseq data. You will see variants that pass all filters, with the 'PASS' keyword in the 7th field of the vcf file (filter status), and also those that has not passed one or few filters.

* *_commonSNPs.diff.sites_in_files_*

Tabular separated file with comparison between variants obtained from RNAseq and DNA data, "known" variants.

The file is sorted by coordinate position and contains 8 fields:

 . CHROM: chromosome name; 
 . POS1: position of the SNV in file #1 (RNAseq data); 
 . POS2: position of SNV in file #2 (DNA "known" variants);
 . IN_FILE: flag whether SNV is present in the file #1 '1', in the file #2 '2', or in both files 'B';
 . REF1: reference sequence in the file 1;
 . REF2: reference sequence in the file 2;
 . ALT1: alternative sequence in the file 1;
 . ALT2: alternative sequence in the file 2

* *_known_snps.vcf_*

Variants that are common between RNAseq and DNA calls, "known" variants.

Allele specific expression quantification::
* *_ASE.tsv_*

Tabular separated file with allele counts at a positions of SNVs (only for common SNVs,those from file known_snps.vcf)
File is sorted by coordinates and contains 13 fields:

. conting: conting,scaffold or chromosome name of the variant
. position: position of the variant
. variant ID: variant ID in the dbSNP
. refAllele: reference allele sequence
. altAllele: alternate allele sequence
. refCount: number of reads that support the reference allele
. altCount: number of reads that support the alternate allele
. totalCount: total number of reads at the site that support both reference and alternate allele and any other alleles present at the site
. lowMAPQDepth: number of reads that have low mapping quality
. lowBaseQDepth: number of reads that have low base quality
. rawDepth: total number of reads at the site that support both reference and alternate allele and any other alleles present at the site
. otherBases: number of reads that support bases other than reference and alternate bases
. improperPairs: number of reads that have malformed pairs


Allele frequency histogram::

* *_AF.histogram.pdf_*

This file contains a histogram plot for allele frequency for SNVs common between RNA-seq and DNA calls, "known" variants

 
