= Worflow Description

The aim of the pipeline is to process raw RNA-seq data (*.fastq) and obtain the list of small variants, SNVs (SNPs and INDELs) for the downsteram analysis. The pipeline is based on the GATK best practices for variant calling with RNAseq data and includes all majour steps. In addition the pipeline includes SNVs postprocessing and quantification for allele specific expression.

_Processing of the samples_ is done for *_each replicate independently_*. This includes mapping of the reads, split at the CIGAR, reassigning mapping qualities and base quality recalibration.  

_The variant calling_ is done on bam files from *_all replicates simultaniously_* – this allows to improve coverage of the genomic regions and perform more reliable variant calling. 

https://software.broadinstitute.org/gatk/guide/article?id=3891[GATK best practices, RNA-seq]

== Sofware manuals

Here are the links for the software documentations.

* http://www.htslib.org/doc/samtools.html[samtools]
* https://broadinstitute.github.io/picard/command-line-overview.html#CreateSequenceDictionary[picard, CreateSequenceDictionary]
* http://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STAR.posix/doc/STARmanual.pdf[STAR]
* https://vcftools.github.io/man_latest.html[vcftools]
* https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/index[GATK tools]
** https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_rnaseq_SplitNCigarReads.php[SplitNCigarReads]
** https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_bqsr_BaseRecalibrator.php[BaseRecalibrator]
** https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_readutils_PrintReads.php[PrintReads]
** https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php[HaplotypeCaller]
** https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_filters_VariantFiltration.php[VariantFiltration]
** https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_rnaseq_ASEReadCounter.php[ASEReadCounter]

== Prepare data

This step prepare input files for the further analysis – genome indexes are created and filter those variants overlapping blacklisted regions. 

Create samtools and pickard genome index  - needed for GATK tools, such as Split'N'Trim

----
samtools faidx genome.fa
java -jar picard.jar CreateSequenceDictionary R= genome.fa O= genome.dict
----

Create STAR 2.5 genome index – needed for RNA-seq reads mappings
STAR creates index file in the folder with name genome_dir
----
STAR --runMode genomeGenerate \
       --genomeDir genome_dir \
       --genomeFastaFiles genome.fa \
       --runThreadN 4
----

Filter variants that overlap with blacklisted regions to reduce false positive calls [optional].

 vcftools --gzvcf known_SNVs.vcf -c \
           --exclude-bed blacklisted.bed \
           --recode | bgzip -c \
           > known_SNVs.filtered.recode.vcf.gz
          
           
== Mapping RNAseq reads to the reference

To align RNA-seq reads to the genome we're using STAR 2-pass approach. The first alignment creates a table with splice-junctions that is used to guide final alignments. The alignments at both steps are done with default parameters. 

Additional fields with the read groups, libraries and sample information are added into the final bam file at the second mapping step. As a result we do not need to run Picard processing step from GATK best practices.

STAR 1-pass:

----
 STAR --genomeDir genome_dir \
       --readFilesIn sample_1.fastq.gz sample_2.fastq.gz \
       --runThreadN 4 \
       --readFilesCommand zcat \
       --outFilterType BySJout \
       --alignSJoverhangMin 8 \
       --alignSJDBoverhangMin 1 \
       --outFilterMismatchNmax 999
----

Create new genome index using splice-junction table:

----
STAR --runMode genomeGenerate \
       --genomeDir genome_dir \
       --genomeFastaFiles genome.fa \
       --sjdbFileChrStartEnd SJ.out.tab \
       --sjdbOverhang 75 \
       --runThreadN 4
----

STAR 2-pass, final alignments:

----
  STAR --genomeDir genome_dir \
       --readFilesIn sample_1.fastq.gz sample_2.fastq.gz \
       --runThreadN 4 \
       --readFilesCommand zcat \
       --outFilterType BySJout \
       --alignSJoverhangMin 8 \
       --alignSJDBoverhangMin 1 \
       --outFilterMismatchNmax 999 \
       --outSAMtype BAM SortedByCoordinate \
       --outSAMattrRGline ID:ReplicateID LB:library PL:illumina PU:machine SM:GM12878
----

Index resulting bam file

----
samtools index final_alignments.bam
----

== Split'N'Trim and reassign mapping qualities

The RNA-seq reads overlapping exon-intron junctions can produce a false positive variants due to inaccurate splicing. To solve this problem the GATK team recommend to hard-clip any sequence that overlap intronic region, and has developed a speciall tool SplitNCigarReads. The tool identify Ns in the CIGAR string of the alignment and split reads at this position, few new reads are created. 

At this step we also reassigning mapping qualities of the alignments. This is important because STAR assign good quality of 255 to the “unknown” mappings  that is meaningless to GATK and to variant calling in general.  

This step is done with recommended parameters from the GATK best practics.

----
java -jar GenomeAnalysisTK.jar -T SplitNCigarReads \
                  -R genome.fa -I final_alignments.bam \
                  -o split.bam \
                  -rf ReassignOneMappingQuality \
                  -RMQF 255 -RMQT 60 \
                  -U ALLOW_N_CIGAR_READS \
                  --fix_misencoded_quality_scores
----

== Base Recalibration

Here in the pipeline we did not included indel realigment step, optional step in the GATK best practices. It is time-intensive but does not really improve variant calling. 

We include base recalibration step – this step allows to remove possible systematic errors introduced by the sequencing machine in the assigned qualities of the reads. To do this, the list of known variants is used as an input for machine learning algorithm that model possible errors. Then the base quality score is adjusted based on the obtained results.

----
java -jar GenomeAnalysisTK.jar -T BaseRecalibrator \
                  --default_platform illumina \
                  -cov ReadGroupCovariate \
                  -cov QualityScoreCovariate \
                  -cov CycleCovariate \
                  -knownSites known_SNVs.filtered.recode.vcf.gz\
                  -cov ContextCovariate \
                  -R genome.fa -I split.bam \
                  --downsampling_type NONE \
                  -nct 4 \
                  -o final.rnaseq.grp 

  java -jar GenomeAnalysisTK.jar -T PrintReads \
                  -R genome.fa -I split.bam \
                  -BQSR final.rnaseq.grp \
                  -nct 4 \
                  -o final.bam

----

== Variant Calling and Variant filtering

The variant calling is done on the uniquelly aligned reads only, it allows to reduce number of false positive variants.

----
(samtools view -H final.bam; samtools view final.bam| grep -w 'NH:i:1') \
  |samtools view -Sb -  > final.uniq.bam
  
samtools index final.uniq.bam
----

For variant calling we're using tool HaplotypeCaller from the GATK tools with default parameters. 

----
ls final.uniq.bam  > bam.list
java -jar GenomeAnalysisTK.jar -T HaplotypeCaller \
                  -R genome.fa -I bam.list \
                  -dontUseSoftClippedBases \
                  -stand_call_conf 20.0 \
                  -o output.gatk.vcf.gz                  
----

Variant filtering is done as recommended by GATK:

* clusters of at least 3 SNPs that are within a window of 35 bases between them 
* strand bias estimated using Fisher's Exact Test with values  > 30.0 (Phred-scaled p-value)
* variant call confidence score QualByDepth (QD) values < 2.0. The QD is the QUAL score normalized by allele depth (AD) for a variant.

----
 java -jar GenomeAnalysisTK.jar -T VariantFiltration \
                  -R genome.fa -V output.gatk.vcf.gz \
                  -window 35 -cluster 3 \
                  -filterName FS -filter "FS > 30.0" \
                  -filterName QD -filter "QD < 2.0" \
                  -o final.vcf
----

== Variant Post-processing

For downsteram analysis we're considering only sites that pass all filters and covered with at least 8 reads.

----
grep -v '#' final.vcf | awk '$7~/PASS/' \
|perl -ne 'chomp($_); ($dp)=$_=~/DP\\=(\\d+)\\;/; if($dp>=8){print $_."\\n"};' > result.DP8.vcf
---- 

Filtered RNA-seq variants are compared with those obtained from DNA sequencing (from Illumina platinum genome project). Those variants that are common between these two datasets, are "known" SNVs, those present in the RNA-seq cohort only are "novel". 

For _allele specific expression_ analysis we will use _only known SNVs_. 
_Novel variants_ will be used to detect _RNA-editing events_.

Compare two variant *.vcf files to detect common and different sites: 
----
 vcftools --vcf result.DP8.vcf --gzdiff known_SNVs.filtered.recode.vcf.gz --diff-site --out commonSNPs
----

Here we select sites present in the both files, "known" SNVs only:

----
awk 'BEGIN{OFS="\t"} $4~/B/{print $1,$2,$3}' commonSNPs.diff.sites_in_files  > test.bed
   
vcftools --vcf final.vcf --bed test.bed --recode --keep-INFO-all --stdout > known_snps.vcf
----

Plot a histogram with allele frequency distribution for "known" SNVs
----
grep -v '#'  known_snps.vcf | awk -F '\\t' '{print $10}' \
               |awk -F ':' '{print $2}'|perl -ne 'chomp($_); \
               @v=split(/\\,/,$_); if($v[0]!=0 ||$v[1] !=0)\
               {print  $v[1]/($v[1]+$v[0])."\\n"; }' |awk '$1!=1' \
               >AF.4R

gghist.R -i AF.4R -o AF.histogram.pdf
----

Calculate read counts per each "known" SNVs per allele for allele specific expression analysis

----
java -jar GenomeAnalysisTK.jar -R genome.fa \
                  -T ASEReadCounter \
                  -o ASE.tsv \
                  -I bam.list \
                  -sites known_snps.vcf
----
