= Part 1

.Complete `main.nf` script for Part 1
----
/*
 * Define the default parameters
 */
params.genome     = "$baseDir/data/genome.fa" 
params.variants   = "$baseDir/data/known_variants.vcf.gz"
params.blacklist  = "$baseDir/data/blacklist.bed"
params.reads      = "$baseDir/data/reads/rep1_{1,2}.fq.gz"
params.results    = "results"
params.gatk       = '/usr/local/bin/GenomeAnalysisTK.jar'


/*
 *  Parse the input parameters
 */
genome_file     =  file(params.genome)
variants_file   =  file(params.variants)
blacklist_file  =  file(params.blacklist)
reads_ch        =  Channel.fromFilePairs(params.reads)
GATK            =  params.gatk


/*
 * Process 1A: Create a FASTA genome index with samtools
 */
process '1A_prepare_genome_samtools' { 
  input:
      file genome from genome_file
  output:
      file "${genome}.fai" into genome_index_ch 
  script:
  """
  samtools faidx ${genome}
  """
}


/*
 * Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK
 */
process '1B_prepare_genome_picard' {
  input:
      file genome from genome_file
  output:
      file "${genome.baseName}.dict" into genome_dict_ch
  script:
  """
  PICARD=`which picard.jar`
  java -jar \$PICARD CreateSequenceDictionary R= $genome O= ${genome.baseName}.dict
  """


/*
 * Process 1C: Create the genome index file for STAR
 */
process '1C_prepare_star_genome_index' {
  input:
      file(genome) from genome_file
  output:
      file(genome_dir) into genome_dir_ch
  script:
  """
  mkdir genome_dir

  STAR --runMode genomeGenerate \
       --genomeDir genome_dir \
       --genomeFastaFiles ${genome} \
       --runThreadN ${task.cpus}


/*
 * Process 1D: Create a file containing the filtered and recoded set of variants
 */

process '1D_prepare_vcf_file' {
  input: 
      file(variantsFile) from variants_file
      file(blacklisted) from blacklist_file
  output:
      set file("${variantsFile.baseName}.filtered.recode.vcf.gz"), \
          file("${variantsFile.baseName}.filtered.recode.vcf.gz.tbi") into prepared_vcf_ch
  script:  
  """
  vcftools --gzvcf $variantsFile -c \
           --exclude-bed ${blacklisted} \
           --recode | bgzip -c \
           > ${variantsFile.baseName}.filtered.recode.vcf.gz

  tabix ${variantsFile.baseName}.filtered.recode.vcf.gz
  """
}


